src/self-update/README.md

```diff
@@ -0,0 +1,66 @@
+* Device Owner shall be able to use a power-fail safe filesystem and transaction-safe package manager (instead of APT) - if the device support any.
+  * Therefore an extensible plugin shall be provided.
+* In thin-edge package's installation logic no external tools (e.g. grep, sed, ...) shall be used. Instead Rust code/libraries shall be used.
+  * This is to reduce external dependencies and so reduce risc for failures due missing tools or wrong version of tools.
+
+# Expected Pre-Conditions 
+
+The device and it's thereon installed thin-edge has to follow precondition below to allow successful self-update:
+
+1) The current (old) thin-edge on the device must be installed with usual thin-edge installation, or at leat as usual thin-edge installation would do. That is to assure new thin-edge version that is about to be installed has all external and internal files, folders and resources (e.g. config files) available.
+2) Deployment of thin-edge and mosquitto must not be changed manually after initial installation (e.g. moving/renaming installed files, moving/renaming configuration files used by thin-edge, renaming users/groups created during installation, ...).
+
+3) Package Dependencies need to be solved using SW Management before self-update. This is since self-update will disable auto-dep solving to reduce additional risk during self-update.
+
+*Hint: Before executing self-update on devices in the field it shall be tested on a kind of golden device that equals the device(s) in the field in system-environment, configuration and thin-edge version.*
+
+
+# Execution Flow and involved Components
+![Sequence Diagram Update SW-list](images/self-update.drawio.svg)
+
+
+# Design Principles
+
+### Run-Time Behaviour
+* Self-Update is provided based on usual thin-egde's Software Management.
+  * A plugin for module type "tedge-update" specifically handles thin-edge self-update.
+* The plugin uses Debian Package Manager APT to upgrade requested thin-edge packages.
+* When executing APT the plugin will disable auto dependency solving.
+  * This is to avoid installing more than the requested packages to reduce risc for potential failures.
+* The plugin shall take all requested packages at once from SM Agent (i.E. Plugin API command "update-list" shall be used).
+* The plugin follows the contract of the Software Management Plugin API, with the exception of the way to report final result to the SM Agent:
+  * As the new started SM Agent will be disconnected from tedge-update plugin's child-process, the plugin will store potential `stdout` and the final exit status to some persistent file.
+  * The SM Agent will pickup that file and generates according update response for the mapper.
+* Potentially the spawned plugin-process need to detouched it-self from parent process SM Agent to avoid any impact when SM Agent process is stopped by APT upgrade procedure.
```

Comment:
I'm not very fond of the whole super special case for updates.
I see a couple options here:
1. Detach the child saving a state (eg. pid, name etc) and after coming back up after the upgrade re attaching to the plugin process and reading stdout/exit codes (we could use `ptrace`). 
  * this would only be compatible with *nix system though
2. Just change slightly how the whole plugin system behaves such that there is always a state file where the plugin dumps results and agent can read them when it is up again. 
  * As much as it requires changes to the agent and the plugin API it will open path for more flexible use of plugins.
  * Requires only small changes to the APT plugin which could then handle upgrade process as proposed above.

Reply:
Now decision was taken to move to MQTT to transfer exit-code from plugin to SM Agent, in case of self-update. Spec is updated accordingly.